services/spar/src/Spar/API.hs

```diff
@@ -0,0 +1,246 @@
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Spar.API
+  ( runServer, app, api
+  , API
+  , APIMeta
+  , APIAuthReq
+  , APIAuthResp
+  , IdpGet
+  , IdpCreate
+  , IdpDelete
+  ) where
+
+import Bilge
+import Control.Monad.Except
+import Data.Maybe (isJust, fromJust)
+import Data.Metrics (metrics)
+import Data.Proxy
+import Data.String.Conversions (ST, cs)
+import Data.String (fromString)
+import "swagger2" Data.Swagger hiding (Header(..))
+  -- NB: this package depends on both types-common, swagger2, so there is no away around this name
+  -- clash other than -XPackageImports.
+import GHC.Stack
+import Lens.Micro
+import Network.HTTP.Client (responseTimeoutMicro)
+import Servant
+import Servant.Swagger
+import Spar.API.Instances ()
+import Spar.App
+import Spar.Options
+import Spar.Types
+import Util.Options (epHost, epPort)
+import Web.Cookie (SetCookie)
+
+import qualified Data.Aeson as Swagger
+import qualified Data.Scientific as Swagger
+import qualified Brig.Types.User as Brig
+import qualified Data.Id as Brig
+import qualified Data.X509 as X509
+import qualified Network.Wai.Handler.Warp as Warp
+import qualified Network.Wai.Utilities.Server as WU
+import qualified SAML2.WebSSO as SAML
+import qualified Servant.Multipart as SM
+import qualified Spar.Data as Data
+import qualified Spar.Intra.Brig as Brig
+import qualified System.Logger as Log
+import qualified URI.ByteString as URI
+
+runServer :: Opts -> IO ()
+runServer sparCtxOpts = do
+  sparCtxLogger <- Log.new $ Log.defSettings
+                   & Log.setLogLevel (toLevel $ saml sparCtxOpts ^. SAML.cfgLogLevel)
+                   & Log.setOutput Log.StdOut
+                   & Log.setFormat Nothing
+                   & Log.setNetStrings (logNetStrings sparCtxOpts)
+  mx <- metrics
+  sparCtxCas <- Data.initCassandra sparCtxOpts sparCtxLogger
+  let settings = Warp.defaultSettings
+        & Warp.setHost (fromString $ sparCtxOpts ^. to saml . SAML.cfgSPHost)
+        . Warp.setPort (sparCtxOpts ^. to saml . SAML.cfgSPPort)
+  sparCtxHttpManager <- newManager defaultManagerSettings
+      { managerResponseTimeout = responseTimeoutMicro (10 * 1000 * 1000)
+      }
+  let sparCtxHttpBrig = Bilge.host (sparCtxOpts ^. to brig . epHost . to cs)
+                      . Bilge.port (sparCtxOpts ^. to brig . epPort)
+                      $ Bilge.empty
+  let wrappedApp
+    -- . WU.measureRequests mx _
+        -- TODO: we need the swagger sitemap from servant for this.  we also want this to be
+        -- prometheus-compatible.  not sure about the order in which to do these.
+        = WU.catchErrors sparCtxLogger mx
+        . SAML.setHttpCachePolicy
+        $ app Env {..}
+  WU.runSettingsWithShutdown settings wrappedApp 5
+
+-- FUTUREWORK: use servant-generic?
+
+app :: Env -> Application
+app ctx = SAML.setHttpCachePolicy
+        $ serve (Proxy @API) (enter (NT (SAML.nt @Spar ctx)) api :: Server API)
+
+type API = "i" :> "status" :> GetNoContent '[JSON] NoContent
```

Comment:
Can we have `/i/status` return `200 OK` instead? 😬

Reply:
b5e5983c3a168564b0532476f06bf057e9e15066

For the record, I still think 200 is wrong here, though:

> 200 OK
> Standard response for successful HTTP requests. The actual response will depend on the request method used. In a GET request, the response will contain an entity corresponding to the requested resource. In a POST request, the response will contain an entity describing or containing the result of the action.[9]

[9] https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#cite_note-RFC_2616-10